### DISCUSSION TITLE
Secure artifacts against modification

### DISCUSSION THREAD

**Discussion by johngmyers** [fluxcd/flux2#1707] [Proposals]
# The Problem

Helm Controller and Kustomize Controller retrieve the artifacts to be installed from Source Controller over an unsecured HTTP connection. An attacker able to inject local network traffic or poison local DNS would be able to escalate their access by injecting malicious content into artifacts being retrieved by the Helm or Kustomize controllers.

# Proposal

Artifacts can be secured against modification by the addition of two features:

1) Add a secure hash to the `Artifact` subresource.

The `Artifact` subresource currently has a `Checksum` field, but that field contains a SHA-1 checksum. SHA-1 is vulnerable to chosen-prefix and length extension attacks and is unsuitable for securing content against malicious modification.

A new `SHA256` field should be added to the `Artifact` subresource, containing a SHA-256 of the artifact. The `Checksum` field can be deprecated.

2) Modify the Helm Controller and Kustomize Controller to verify the SHA-256 of the content they download from source controller matches the `SHA256` field of the corresponding `Artifact`. If it does not match, the controllers should not install or use the downloaded content.

**Comment by hiddeco** [comment]
This was already on the unofficial planning! But I sped things a bit up by already creating https://github.com/fluxcd/source-controller/pull/423 (NB: the target branch of this PR is a development branch with a much bigger unit of work, and it will take some time before this lands in an official release).

**Reply by HenryHoggard** [reply]
@hiddeco , @johngmyers, how would sha256 protect against (malicious) modification of the artifact if the hash is colocated with the artifact? If we're assuming breach of the artifact host, or MitM, surely they would just modify the checksum too?

Have you considered using signature based verification?

Potentially using something like:
https://www.sigstore.dev/ 
or
https://github.com/notaryproject/nv2

**Reply by johngmyers** [reply]
The hash is transmitted as part of the custom resource from the source controller to the Kubernetes API server and thence to the Helm and/or Kustomize controller. All of those transmission paths are secured against MitM with HTTPS using the cluster's PKI.

**Reply by hiddeco** [reply]
> If we're assuming breach of the artifact host, [..]

If we are assuming a total breach of the artifact host (== source-controller container), most hope is lost at present, as this will grant the attacker access to both the storage and Kubernetes API to write whatever truth they need.

> Have you considered using signature based verification? [..]

Sigstore is relatively new but already on our radar, given this, there will likely be more improvements related to source and/or file integrity in the future.
